File: go/flocking.go
// "Go to the Ant" — a faithful Go port of Reynolds' boids flocking.
//
// Source: H. Van Dyke Parunak, "'Go to the Ant': Engineering Principles from
// Natural Multi-Agent Systems," Annals of Operations Research 75:69-101 (1997),
// §3.5 (Birds & Fish: Flocking), after Reynolds 1987 and Heppner 1990.
//
// Ported from the authoritative Python reference (flocking.py). Behavior is
// reproduced structurally; numbers differ from CPython only because this port
// uses a shared SplitMix64 PRNG (identical across all language ports) rather
// than Python's Mersenne Twister. The five sequential ports agree with EACH
// OTHER.
//
// THE LENS (Go): each boid is an AGENT (a struct index into shared slices) and
// the flock is the SHARED STORE all agents step over. This is a step toward
// goroutine-per-agent — here the agents still step sequentially in index order
// over one mutable store, which keeps the update deterministic and identical to
// the other sequential ports. The actor/concurrency framing lives in the naming
// (Agent, shared Flock), not yet in real parallelism.
//
// Reynolds' three local rules (§3.5), preserved as provenance — these ARE the
// paper's rules:
//  1. SEPARATION — keep a minimum distance from the nearest birds (avoid collisions).
//  2. ALIGNMENT  — match velocity (speed + heading) to nearby birds.
//  3. COHESION   — stay close to the centre of the local flock.
//
// Each rule is a steering vector from the neighbours inside a perception radius;
// their weighted sum turns the bird. A single coherent, banking flock EMERGES —
// no leader, no central coordinator; each bird senses only its nearest peers.
//
// PROVENANCE: the three rules are the paper's (Reynolds' "boids"). The perception
// radius, the separation distance, and the three weights are OPERATIONALIZED —
// Parunak lists the rules but gives no numbers (Reynolds 1987 is the primary
// source for tuned constants). NO per-step randomness: the run is fully
// deterministic given the random init, so cross-port identity depends only on
// matching the init-RNG order and the neighbour-sum order (iterate j in index
// order).
package main

import (
	"flag"
	"fmt"
	"math"
)

// ---- SplitMix64 PRNG (shared across all ports; standard library only) --------

type SplitMix64 struct{ state uint64 }

func (r *SplitMix64) next() uint64 {
	r.state += 0x9E3779B97F4A7C15
	z := r.state
	z = (z ^ (z >> 30)) * 0xBF58476D1CE4E5B9
	z = (z ^ (z >> 27)) * 0x94D049BB133111EB
	return z ^ (z >> 31)
}

// randomFloat in [0,1): (next() >> 11) * 2^-53
func (r *SplitMix64) randomFloat() float64 {
	return float64(r.next()>>11) * (1.0 / 9007199254740992.0)
}

// uniform(a,b): a + random_float()*(b-a)
func (r *SplitMix64) uniform(a, b float64) float64 {
	return a + r.randomFloat()*(b-a)
}

// ---- The shared store: Flock holds the boids' position/velocity slices --------

type Flock struct {
	n, w, h                      int
	px, py, vx, vy               []float64
	perc, sepR                   float64
	wsep, wali, wcoh, vmax, turn float64
}

func NewFlock(n, w, h, seed int) *Flock {
	fl := &Flock{
		n: n, w: w, h: h,
		perc: 8.0, sepR: 3.0, // OPERATIONALIZED (Reynolds 1987 tuned constants)
		wsep: 1.3, wali: 1.5, wcoh: 0.85, // OPERATIONALIZED rule weights
		vmax: 1.0, turn: 0.35,
	}
	rng := &SplitMix64{state: uint64(seed)}
	// RNG consume order MUST match the Python reference: all px, then all py,
	// then all headings. uniform(0,w) / uniform(0,h) / uniform(0,2pi).
	fl.px = make([]float64, n)
	fl.py = make([]float64, n)
	fl.vx = make([]float64, n)
	fl.vy = make([]float64, n)
	for i := 0; i < n; i++ {
		fl.px[i] = rng.uniform(0, float64(w))
	}
	for i := 0; i < n; i++ {
		fl.py[i] = rng.uniform(0, float64(h))
	}
	for i := 0; i < n; i++ {
		a := rng.uniform(0, 2*math.Pi)
		fl.vx[i] = math.Cos(a) // vx = cos(heading)
		fl.vy[i] = math.Sin(a) // vy = sin(heading)
	}
	return fl
}

// unit: normalize (x,y) to a unit vector; (0,0) below the tolerance.
func unit(x, y float64) (float64, float64) {
	m := math.Hypot(x, y)
	if m > 1e-9 {
		return x / m, y / m
	}
	return 0.0, 0.0
}

func (fl *Flock) step() float64 {
	n := fl.n
	// The three urges are computed against the CURRENT store, written into a
	// scratch velocity, then committed simultaneously (Python does nvx=vx[:]).
	nvx := make([]float64, n)
	nvy := make([]float64, n)
	copy(nvx, fl.vx)
	copy(nvy, fl.vy)
	p2 := fl.perc * fl.perc
	s2 := fl.sepR * fl.sepR
	w := float64(fl.w)
	h := float64(fl.h)
	for i := 0; i < n; i++ {
		var sx, sy, ax, ay, cx, cy float64
		cnt := 0
		for j := 0; j < n; j++ { // iterate j in index order (cross-port identity)
			if i == j {
				continue
			}
			dx := fl.px[j] - fl.px[i]
			dy := fl.py[j] - fl.py[i]
			dx -= w * math.Round(dx/w) // toroidal delta
			dy -= h * math.Round(dy/h)
			d2 := dx*dx + dy*dy
			if d2 > p2 {
				continue
			}
			cnt++
			ax += fl.vx[j] // rule 2: alignment (avg neighbour velocity)
			ay += fl.vy[j]
			cx += dx // rule 3: cohesion (toward neighbour centre)
			cy += dy
			if d2 < s2 && d2 > 1e-9 { // rule 1: separation (push from the close ones)
				sx -= dx / d2
				sy -= dy / d2
			}
		}
		if cnt > 0 {
			fc := float64(cnt)
			ax /= fc
			ay /= fc
			cx /= fc
			cy /= fc
			// NORMALIZE each urge to a unit vector so the three weights are
			// actually comparable (otherwise the position-scale cohesion vector
			// swamps the velocity-scale alignment one).
			sux, suy := unit(sx, sy)                   // rule 1: away from close birds
			aux, auy := unit(ax-fl.vx[i], ay-fl.vy[i]) // rule 2: toward neighbours' heading
			cux, cuy := unit(cx, cy)                   // rule 3: toward neighbours' centre
			accx := fl.wsep*sux + fl.wali*aux + fl.wcoh*cux
			accy := fl.wsep*suy + fl.wali*auy + fl.wcoh*cuy
			nvx[i] = fl.vx[i] + fl.turn*accx
			nvy[i] = fl.vy[i] + fl.turn*accy
			sp := math.Hypot(nvx[i], nvy[i]) // cap speed
			if sp == 0 {
				sp = 1.0
			}
			nvx[i] = nvx[i] / sp * fl.vmax
			nvy[i] = nvy[i] / sp * fl.vmax
		}
	}
	// commit: update all velocities then all positions toroidally.
	for i := 0; i < n; i++ {
		fl.vx[i] = nvx[i]
		fl.vy[i] = nvy[i]
		fl.px[i] = math.Mod(fl.px[i]+fl.vx[i], w)
		if fl.px[i] < 0 {
			fl.px[i] += w
		}
		fl.py[i] = math.Mod(fl.py[i]+fl.vy[i], h)
		if fl.py[i] < 0 {
			fl.py[i] += h
		}
	}
	return fl.polarization()
}

// polarization: order parameter |mean heading| / vmax. 0 = disordered chaos,
// 1 = one coherent flock all pointing the same way.
func (fl *Flock) polarization() float64 {
	var mx, my float64
	for i := 0; i < fl.n; i++ {
		mx += fl.vx[i]
		my += fl.vy[i]
	}
	mx /= float64(fl.n)
	my /= float64(fl.n)
	return math.Hypot(mx, my) / fl.vmax
}

// ---- Driver ------------------------------------------------------------------

func run(ticks, n, seed int) (*Flock, []float64) {
	fl := NewFlock(n, 90, 48, seed)
	hist := []float64{fl.polarization()}
	step := ticks / 12
	if step < 1 {
		step = 1
	}
	for t := 0; t < ticks; t++ {
		p := fl.step()
		if t%step == 0 {
			hist = append(hist, p)
		}
	}
	return fl, hist
}

func render(fl *Flock) {
	// arrow[k] for heading bucket k = round(atan2/(pi/4)) mod 8.
	arrow := []rune{'→', '↗', '↑', '↖', '←', '↙', '↓', '↘'}
	// one rune per grid cell (arrows are multibyte); build lines as strings.
	cell := make([][]rune, fl.h)
	for y := range cell {
		cell[y] = make([]rune, fl.w)
		for x := range cell[y] {
			cell[y][x] = ' '
		}
	}
	for i := 0; i < fl.n; i++ {
		x := ((int(fl.px[i]) % fl.w) + fl.w) % fl.w
		y := ((int(fl.py[i]) % fl.h) + fl.h) % fl.h
		a := math.Atan2(fl.vy[i], fl.vx[i])
		k := int(math.Round(a/(math.Pi/4))) % 8
		if k < 0 {
			k += 8
		}
		cell[y][x] = arrow[k]
	}
	fmt.Print("\nFlock (each bird points along its heading — watch them align):\n\n")
	for y := 0; y < fl.h; y++ {
		fmt.Println(string(cell[y]))
	}
}

func main() {
	ticks := flag.Int("ticks", 600, "number of ticks")
	birds := flag.Int("birds", 90, "number of boids")
	ants := flag.Int("ants", 0, "alias for --birds (0 = use --birds)")
	seed := flag.Int("seed", 0, "PRNG seed")
	flag.Parse()

	n := *birds
	if *ants != 0 {
		n = *ants
	}

	fl, hist := run(*ticks, n, *seed)
	render(fl)
	fmt.Printf("\npolarization (flock alignment): %.3f  (0 = chaos, 1 = one flock)\n", fl.polarization())
	fmt.Print("polarization(t):")
	for _, c := range hist {
		fmt.Printf(" %.2f", c)
	}
	fmt.Println()
}

File: java/Flocking.java
/*
 * "Go to the Ant" §3.5 — Birds & Fish: Flocking (Reynolds 1987, Heppner 1990),
 * a faithful Java port of the reference recreated from Parunak's survey.
 *
 * Source: H. Van Dyke Parunak, "'Go to the Ant': Engineering Principles from Natural
 * Multi-Agent Systems," Annals of Operations Research 75:69-101 (1997).
 *
 * LENS: Java — the classical JVM agent-based-modeling lineage (MASON, Repast, NetLogo).
 * The idiom is explicit objects under a schedule: a World that OWNS the shared state
 * arrays (px, py, vx, vy) and a swarm of individually-instantiated Boid objects, each
 * carrying only its own index and the three local steering rules. The World steps every
 * Boid to compute a next velocity, THEN commits positions — a two-phase schedule so no
 * boid sees a half-updated flock. There is no leader and no central plan.
 *
 * Reynolds' three local rules (§3.5), which are the paper's:
 *   1. SEPARATION — keep a minimum distance from the nearest birds (avoid collisions).
 *   2. ALIGNMENT  — match velocity (speed + heading) to nearby birds.
 *   3. COHESION   — steer toward the centre of the local flock.
 * A single coherent, banking flock EMERGES from these three local urges.
 *
 * PROVENANCE: the three rules are the paper's (Reynolds' "boids"). The perception radius,
 * the separation distance, and the three weights are OPERATIONALIZED — Parunak's paper
 * lists the rules but gives no numbers (Reynolds 1987 is the primary source for tuned
 * constants). NO per-step randomness: the run is fully deterministic given the random
 * init, so cross-port identity depends only on matching the init-RNG order and the
 * neighbour-sum order (iterate j in index order).
 *
 * DEPENDENCY-FREE: standard library only (RNG is a hand-rolled SplitMix64).
 */

public class Flocking {

    // ---- SplitMix64 PRNG: identical across all ports so they are directly comparable. ----
    static final class SplitMix64 {
        private long state;
        SplitMix64(long seed) { this.state = seed; }
        long next() {
            state += 0x9E3779B97F4A7C15L;
            long z = state;
            z = (z ^ (z >>> 30)) * 0xBF58476D1CE4E5B9L;
            z = (z ^ (z >>> 27)) * 0x94D049BB133111EBL;
            return z ^ (z >>> 31);
        }
        // random_float() in [0,1): (next() >> 11) * 2^-53
        double random() {
            return (next() >>> 11) * (1.0 / 9007199254740992.0);
        }
        // uniform(a,b): a + random_float()*(b-a)
        double uniform(double a, double b) {
            return a + random() * (b - a);
        }
    }

    // The World owns the shared flock state; Boids read/write into it under the schedule.
    static final class World {
        final int n, w, h;
        final SplitMix64 rng;
        final double[] px, py, vx, vy;   // shared arrays: position and (unit-speed) velocity
        final double[] nvx, nvy;         // scratch for the two-phase (compute-then-commit) step
        final double perc, sepR, wsep, wali, wcoh, vmax, turn;

        World(int n, int w, int h, long seed,
              double perc, double sepR, double wsep, double wali, double wcoh,
              double vmax, double turn) {
            this.n = n; this.w = w; this.h = h;
            this.rng = new SplitMix64(seed);
            this.perc = perc; this.sepR = sepR;
            this.wsep = wsep; this.wali = wali; this.wcoh = wcoh;
            this.vmax = vmax; this.turn = turn;
            this.px = new double[n]; this.py = new double[n];
            this.vx = new double[n]; this.vy = new double[n];
            this.nvx = new double[n]; this.nvy = new double[n];
            // Init: px,py uniform in the box; heading uniform; vx=cos, vy=sin.
            // RNG consumption order MUST match the Python reference exactly:
            // all px, then all py, then all angles.
            for (int i = 0; i < n; i++) px[i] = rng.uniform(0, w);
            for (int i = 0; i < n; i++) py[i] = rng.uniform(0, h);
            for (int i = 0; i < n; i++) {
                double a = rng.uniform(0, 2 * Math.PI);
                vx[i] = Math.cos(a); vy[i] = Math.sin(a);
            }
        }

        // Order parameter: |mean heading| / vmax. 0 = disordered, 1 = one coherent flock.
        double polarization() {
            double mx = 0, my = 0;
            for (int i = 0; i < n; i++) { mx += vx[i]; my += vy[i]; }
            mx /= n; my /= n;
            return Math.hypot(mx, my) / vmax;
        }

        // Two-phase schedule: every Boid computes its next velocity into (nvx,nvy) from the
        // frozen current state, THEN we commit velocities and move positions toroidally.
        double step(Boid[] boids) {
            for (int i = 0; i < n; i++) { nvx[i] = vx[i]; nvy[i] = vy[i]; }
            for (Boid b : boids) b.computeNextVelocity();
            for (int i = 0; i < n; i++) {
                vx[i] = nvx[i]; vy[i] = nvy[i];
                px[i] = mod(px[i] + vx[i], w);
                py[i] = mod(py[i] + vy[i], h);
            }
            return polarization();
        }

        // Python's % on floats always returns a result with the divisor's sign (non-negative here).
        static double mod(double a, double m) {
            double r = a % m;
            return r < 0 ? r + m : r;
        }
    }

    // A single bird. Holds only its index and a back-reference to the shared World — all
    // steering is LOCAL (neighbours inside the perception radius).
    static final class Boid {
        final World world;
        final int i;
        Boid(World world, int i) { this.world = world; this.i = i; }

        void computeNextVelocity() {
            final World wd = world;
            final int n = wd.n;
            final double[] px = wd.px, py = wd.py, vx = wd.vx, vy = wd.vy;
            final double p2 = wd.perc * wd.perc, s2 = wd.sepR * wd.sepR;
            double sx = 0, sy = 0, ax = 0, ay = 0, cx = 0, cy = 0;
            int cnt = 0;
            // Iterate j in index order to keep the neighbour-sum bit-identical across ports.
            for (int j = 0; j < n; j++) {
                if (i == j) continue;
                double dx = px[j] - px[i], dy = py[j] - py[i];
                // toroidal delta (VERBATIM: dx -= w*round(dx/w)); Math.rint == Python round (half-to-even)
                dx -= wd.w * Math.rint(dx / wd.w);
                dy -= wd.h * Math.rint(dy / wd.h);
                double d2 = dx * dx + dy * dy;
                if (d2 > p2) continue;
                cnt++;
                ax += vx[j]; ay += vy[j];          // rule 2: alignment (avg neighbour velocity)
                cx += dx; cy += dy;                // rule 3: cohesion (toward neighbour centre)
                if (d2 < s2 && d2 > 1e-9) {         // rule 1: separation (push from the close ones)
                    sx -= dx / d2; sy -= dy / d2;
                }
            }
            if (cnt == 0) return;                   // no neighbours: keep current velocity
            ax /= cnt; ay /= cnt; cx /= cnt; cy /= cnt;
            // NORMALIZE each urge to a unit vector so the three weights are actually comparable
            // (otherwise the position-scale cohesion vector swamps the velocity-scale alignment one).
            double sux = 0, suy = 0, aux = 0, auy = 0, cux = 0, cuy = 0;
            double m;
            m = Math.hypot(sx, sy);              if (m > 1e-9) { sux = sx / m; suy = sy / m; }
            double avx = ax - vx[i], avy = ay - vy[i];
            m = Math.hypot(avx, avy);            if (m > 1e-9) { aux = avx / m; auy = avy / m; }
            m = Math.hypot(cx, cy);             if (m > 1e-9) { cux = cx / m; cuy = cy / m; }
            double accx = wd.wsep * sux + wd.wali * aux + wd.wcoh * cux;
            double accy = wd.wsep * suy + wd.wali * auy + wd.wcoh * cuy;
            double nvx = vx[i] + wd.turn * accx, nvy = vy[i] + wd.turn * accy;
            double sp = Math.hypot(nvx, nvy);
            if (sp == 0.0) sp = 1.0;                // cap speed to vmax
            wd.nvx[i] = nvx / sp * wd.vmax;
            wd.nvy[i] = nvy / sp * wd.vmax;
        }
    }

    static void render(World wd) {
        char[] arrow = {'→', '↗', '↑', '↖', '←', '↙', '↓', '↘'};
        char[][] grid = new char[wd.h][wd.w];
        for (int y = 0; y < wd.h; y++) for (int x = 0; x < wd.w; x++) grid[y][x] = ' ';
        for (int i = 0; i < wd.n; i++) {
            int x = ((int) wd.px[i] % wd.w + wd.w) % wd.w;
            int y = ((int) wd.py[i] % wd.h + wd.h) % wd.h;
            double a = Math.atan2(wd.vy[i], wd.vx[i]);
            int k = ((int) Math.rint(a / (Math.PI / 4)) % 8 + 8) % 8;
            grid[y][x] = arrow[k];
        }
        System.out.println();
        System.out.println("Flock (each bird points along its heading — watch them align):");
        System.out.println();
        StringBuilder out = new StringBuilder();
        for (int y = 0; y < wd.h; y++) {
            out.append(new String(grid[y])).append('\n');
        }
        System.out.print(out);
    }

    static void run(int ticks, int n, long seed) {
        // n=90; world 90x48; perc=8; sep_r=3; wsep=1.3; wali=1.5; wcoh=0.85; vmax=1.0; turn=0.35.
        World wd = new World(n, 90, 48, seed, 8.0, 3.0, 1.3, 1.5, 0.85, 1.0, 0.35);
        Boid[] boids = new Boid[n];
        for (int i = 0; i < n; i++) boids[i] = new Boid(wd, i);

        StringBuilder hist = new StringBuilder();
        hist.append(String.format("%.2f", wd.polarization()));
        int sample = Math.max(1, ticks / 12);
        for (int t = 0; t < ticks; t++) {
            double p = wd.step(boids);
            if (t % sample == 0) {
                hist.append(' ').append(String.format("%.2f", p));
            }
        }
        render(wd);
        System.out.println();
        System.out.printf("polarization (flock alignment): %.3f  (0 = chaos, 1 = one flock)%n",
                wd.polarization());
        System.out.println("polarization(t): " + hist);
    }

    public static void main(String[] args) {
        int ticks = 600, birds = 90;
        long seed = 0;
        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "--ticks": ticks = Integer.parseInt(args[++i]); break;
                case "--birds": birds = Integer.parseInt(args[++i]); break;
                case "--ants":  birds = Integer.parseInt(args[++i]); break;  // alias for parity
                case "--seed":  seed  = Long.parseLong(args[++i]); break;
                default: /* ignore unknown args */ break;
            }
        }
        run(ticks, birds, seed);
    }
}
